# <CENTER>DATA ANALYTICS LABORATORY</CENTER>

LIST OF EXPERIMENTS:
1. Random Sampling
2. Z-test case study
3. T-test case studies
4. ANOVA case studies
5. Regression
6. Logistic Regression
7. Time series Analysis

# RANDOM SAMPLING

In [1]:
age_list=[29, 18, 20, 22, 19, 25, 30, 28,22, 21, 18, 19, 20, 20, 22, 23]

In [2]:
import random 
print("First random choice:", random.choice(age_list))
print("Second random choice:", random.choice(age_list))
print("Third random choice:", random.choice(age_list))

First random choice: 19
Second random choice: 30
Third random choice: 28


In [3]:
# Picking Random Items in a List using ‘random.sample()’
print("Random sample, N = 5 :", random.sample(age_list, 5))

Random sample, N = 5 : [22, 19, 20, 22, 29]


In [4]:
# Randomly Shuffling Items in a List using ‘random.shuffle()’
print("AGE list: ", age_list)
random.shuffle(age_list)
print("Shuffled age list: ", age_list)

AGE list:  [29, 18, 20, 22, 19, 25, 30, 28, 22, 21, 18, 19, 20, 20, 22, 23]
Shuffled age list:  [19, 22, 18, 19, 23, 22, 21, 28, 20, 20, 22, 25, 20, 29, 30, 18]


In [5]:
# Generating Random Integers using ‘random.randint()’
print("Random Integer: ", random.randint(1,5))

Random Integer:  2


In [6]:
# Using this function, we can easily generate a list of random integers in a for-loop:
random_ints_list = []
for i in range(1,50):
    n = random.randint(1,5)
    random_ints_list.append(n)
print("My random integer list: ", random_ints_list)

My random integer list:  [3, 5, 2, 4, 5, 4, 5, 2, 4, 1, 3, 2, 5, 4, 2, 5, 4, 5, 1, 2, 2, 4, 3, 4, 5, 1, 5, 1, 5, 3, 2, 3, 3, 1, 3, 3, 3, 5, 1, 5, 3, 3, 2, 5, 4, 5, 4, 4, 1]


In [7]:
#Generating Random Floating Point Values
print("Random Float: ", random.random())

Random Float:  0.6403850612742055


In [8]:
# Generate a list of random floats between 0 and 1:
random_float_list = []
for i in range(1,5):
    n = random.random()
    random_float_list.append(n)
print("My random float list: ", random_float_list)

My random float list:  [0.19760029007042446, 0.6286242291432685, 0.12774938019736037, 0.13911280435982476]


In [2]:
#Computing Uniformly Distributed Numbers with ‘random.uniform()’
import numpy as np
uniform_list = np.random.uniform(-10,1,50)
print("Uniformly Distributed Numbers: ", uniform_list)

Uniformly Distributed Numbers:  [-3.10040455 -0.60528096 -3.78257237 -0.15913439 -7.05614073 -9.12520435
 -3.04253879 -7.31372043 -7.20714528 -0.82581134 -3.35237234 -9.67006248
 -1.44408075 -1.16569778 -5.60797072 -0.97522333 -9.37175466  0.62106767
 -8.2975512   0.33837829  0.4137448   0.32738631  0.47719856 -2.76151533
 -3.99203285 -1.33133542 -8.01877329 -9.56356922 -8.90683304 -7.31259154
 -4.3627132  -6.50186361 -7.2709674  -4.09580322 -7.2646906  -5.13882368
 -1.06424123 -6.84942137 -1.81084055 -6.92717963 -9.21419455 -1.76124222
 -6.5097154   0.33489201 -3.9029948  -7.44117151 -4.5830186  -0.4712773
 -2.27050697 -6.61954658]


In [3]:
# Computing Normally Distributed Numbers with ‘random.gauss()’
normal_list = np.random.uniform(-50,0,50)
print("Normally Distributed Numbers: ", normal_list)

Normally Distributed Numbers:  [-10.19721259 -17.84025245  -8.60752213 -27.52246302  -3.4962885
 -32.29619086  -4.75822213 -21.88809751 -31.16624201  -9.62324347
 -43.76158182 -21.72381849 -44.55633403 -44.70884852 -49.65438593
 -49.81868696 -49.85152977 -34.2060938  -45.70898505 -17.25359332
  -5.67096774 -23.0761587   -9.09896845  -3.40997564 -10.88292299
 -18.75766928  -1.9439984  -31.56390978  -0.67555135 -30.86906695
 -21.62333142 -30.79243964 -16.98499367 -36.8561205  -41.53173531
 -11.98908785 -46.24211502  -7.47407403 -43.31721037 -47.69960757
 -48.30123408  -2.30289114 -43.28673098 -26.32469623  -3.10719097
  -2.40455165 -44.38641671 -47.34214835 -38.90235377 -27.96398727]


# Z-Test

In [11]:
# One Sample Z-Test in Python

<p>Suppose the IQ in a certain population is normally distributed with a mean of μ = 100 and standard deviation of σ = 15.

A researcher wants to know if a new drug affects IQ levels, so he recruits 20 patients to try it and records their IQ levels.

The following code shows how to perform a one sample z-test in Python to determine if the new drug causes a significant difference in IQ levels:
    </p>

In [12]:
from statsmodels.stats.weightstats import ztest as ztest
#enter IQ levels for 20 patients
data = [88, 92, 94, 94, 96, 97, 97, 97, 99, 99,105, 109, 109, 109, 110, 112, 112, 113, 114, 115]
#perform one sample z-test
ztest(data, value=100)

C:\Users\ROHINI\Anaconda3\lib\site-packages\pandas\compat\_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
  warnings.warn(msg, UserWarning)
C:\Users\ROHINI\Anaconda3\lib\site-packages\statsmodels\tools\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


(1.5976240527147705, 0.1101266701438426)

<p>The test statistic for the one sample z-test is 1.5976 and the corresponding p-value is 0.1101. Since this p-value is not less than .05, we do not have sufficient evidence to reject the null hypothesis. In other words, the new drug does not significantly affect IQ level.</p>

In [13]:
# Two Sample Z-Test in Python

<p> Suppose the IQ levels among individuals in two different cities are known to be normally distributed with known standard deviations.A researcher wants to know if the mean IQ level between individuals in city A and city B are different, so she selects a simple random sample of  20 individuals from each city and records their IQ levels.The following code shows how to perform a two sample z-test in Python to determine if the mean IQ level is different between the two cities:</p>

In [14]:
from statsmodels.stats.weightstats import ztest as ztest

#enter IQ levels for 20 individuals from each city
cityA = [82, 84, 85, 89, 91, 91, 92, 94, 99, 99,105, 109, 109, 109, 110, 112, 112, 113, 114, 114]
cityB = [90, 91, 91, 91, 95, 95, 99, 99, 108, 109,109, 114, 115, 116, 117, 117, 128, 129, 130, 133]
#perform two sample z-test
ztest(cityA, cityB, value=0) 

(-1.9953236073282115, 0.046007596761332065)

<p> The test statistic for the two sample z-test is -1.9953 and the corresponding p-value is 0.0460.Since this p-value is less than .05, we have sufficient evidence to reject the null hypothesis. In other words, the mean IQ level is significantly different between the two cities.</p>

# T-Test

T-Test is used to evaluate the means of one or two populations using hypothesis testing.

<p> Systolic blood pressures of 14 patients are given below: 183, 152, 178, 157, 194, 163, 144, 114, 178, 152, 118, 158, 172, 138

Test, whether the population mean, is less than 165</p>
<p>Hypothesis</p>
<p> H0: There is no significant mean difference in systolic blood pressure. i.e., μ = 165 </p>
<p> H1: The population mean is less than 165. i.e., μ < 165 </p>

<br>Test statistic  t= (x̄-μ)/(s/sqrt(n))
<br> Where,
<br> x̄ is sample mean
<br> μ is the population mean
<br> s is sample standard deviation
<br> n is the number of observations


In [17]:
bp=[183, 152, 178, 157, 194, 163, 144, 114, 178, 152, 118, 158, 172, 138]
mu=165
from scipy import stats
t_value,p_value=stats.ttest_1samp(bp,mu)
one_tailed_p_value=float("{:.6f}".format(p_value/2)) # Since alternative hypothesis is one tailed, We need to divide the p value by 2.
print('Test statistic is %f'%float("{:.6f}".format(t_value)))
print('p-value for one tailed test is %f'%one_tailed_p_value)
alpha = 0.05
if one_tailed_p_value<=alpha:
    print('Conclusion','n','Since p-value(=%f)'%p_value,'<','alpha(=%.2f)'%alpha,'''We reject the null hypothesis H0. So we conclude that there is no significant mean difference in systolic blood pressure. i.e., μ = 165 at %.2f level of significance'''%alpha)
else:
    print('Conclusion','n','Since p-value(=%f)'%one_tailed_p_value,'>','alpha(=%.2f)'%alpha,"We do not reject the null hypothesis H0.")
    

Test statistic is -1.243183
p-value for one tailed test is 0.117877
Conclusion n Since p-value(=0.117877) > alpha(=0.05) We do not reject the null hypothesis H0.


<p> So we conclude that there is a significant mean difference in systolic blood pressure. </p>

# ANOVA Test

In [20]:
# Perform a One-Way ANOVA in Python

<p> One-Way ANOVA in Python: One-way ANOVA (also known as “analysis of variance”) is a test that is used to find out whether there exists a statistically significant difference between the mean values of more than one group. </p>

Hypothesis involved:
<p> A one-way ANOVA has the below given null and alternative hypotheses: </p>
<p> H0 (null hypothesis): μ1 = μ2 = μ3 = … = μk (It implies that the means of all the population are equal) </p>
<p> H1 (alternative hypothesis): It states that there will be at least one population mean that differs from the rest </p>

<p> Researchers took 20 cars of the same to take part in a study. These cars are randomly doped with one of the four-engine oils and allowed to run freely for 100 kilometers each. At the end of the journey, the performance of each of the cars is noted. </p>

In [21]:
# Performance when each of the engine oil is applied
performance1 = [89, 89, 88, 78, 79]
performance2 = [93, 92, 94, 89, 88]
performance3 = [89, 88, 89, 93, 90]
performance4 = [81, 78, 81, 92, 82]

In [22]:
from scipy.stats import f_oneway
# Performance when each of the engine oil is applied
performance1 = [89, 89, 88, 78, 79]
performance2 = [93, 92, 94, 89, 88]
performance3 = [89, 88, 89, 93, 90]
performance4 = [81, 78, 81, 92, 82]
# Conduct the one-way ANOVA
f_oneway(performance1, performance2, performance3, performance4)

F_onewayResult(statistic=4.625000000000002, pvalue=0.016336459839780215)

<p> The F statistic and p-value turn out to be equal to 4.625 and 0.016336498 respectively. Since the p-value is lesser than 0.05 hence we reject the null hypothesis. This implies that we have sufficient proof to say that there exists a difference in the performance among four different engine oils. </p>

In [ ]:
# Perform 2 Way ANOVA Test

<p>A botanist wants to know whether or not plant growth is influenced by sunlight exposure and watering frequency. She plants 30 seeds and lets them grow for two months under different conditions for sunlight exposure and watering frequency. After two months, she records the height of each plant, in inches.

Use the following steps to perform a two-way ANOVA to determine if watering frequency and sunlight exposure have a significant effect on plant growth, and to determine if there is any interaction effect between watering frequency and sunlight exposure.</p>

In [8]:
import numpy as np
import pandas as pd

#create data
df = pd.DataFrame({'water': np.repeat(['daily', 'weekly'], 15),
                   'sun': np.tile(np.repeat(['low', 'med', 'high'], 5), 2),
                   'height': [6, 6, 6, 5, 6, 5, 5, 6, 4, 5,
                              6, 6, 7, 8, 7, 3, 4, 4, 4, 5,
                              4, 4, 4, 4, 4, 5, 6, 6, 7, 8]})

#view first ten rows of data 
df[:10]

water  sun  height
0  daily  low       6
1  daily  low       6
2  daily  low       6
3  daily  low       5
4  daily  low       6
5  daily  med       5
6  daily  med       5
7  daily  med       6
8  daily  med       4
9  daily  med       5

In [9]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

#perform two-way ANOVA
model = ols('height ~ C(water) + C(sun) + C(water):C(sun)', data=df).fit()
sm.stats.anova_lm(model, typ=2)

AttributeError: module 'pandas' has no attribute 'Panel'

<p>We can see the following p-values for each of the factors in the table:

water: p-value = .000527
sun: p-value = .0000002
water*sun: p-value = .120667
Since the p-values for water and sun are both less than .05, this means that both factors have a statistically significant effect on plant height. </p>

# Linear Regression

In [23]:
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
exp=[3,8,9,13,3,6]
salary=[30,57,64,72,36,46]
X=np.array([exp]).T
Y=np.array(salary)
X_new=np.array([5,6,7,8])
r=LinearRegression()
r=r.fit(X,Y)
#print(r.coef_)
#print(r.intercept_)
predict_Y=r.predict(X)
print(predict_Y)
print('meansquare',mean_squared_error(Y,predict_Y))
print('RMSE',np.sqrt(mean_squared_error(Y,predict_Y)))

[34.23873874 54.98198198 59.13063063 75.72522523 34.23873874 46.68468468]
meansquare 10.533033033033037
RMSE 3.245463454274757
